File: ga/glm4_result_2024_6/result/1968F_glm/10_1968F.c
use std::io::{self, BufRead};

fn main() {
    let stdin = io::stdin();
    let mut lines = stdin.lock().lines();

    let t: usize = lines.next().unwrap().unwrap().parse().unwrap();

    while t > 0 {
        let n: usize = lines.next().unwrap().unwrap().parse().unwrap();
        let q: usize = lines.next().unwrap().unwrap().parse().unwrap();

        let mut aa: Vec<i32> = (0..=n).map(|_| lines.next().unwrap().unwrap().parse().unwrap()).collect();
        for i in 1..=n {
            aa[i] ^= aa[i - 1];
        }

        let mut ii: Vec<usize> = (0..=n).collect();

        fn sort(aa: &Vec<i32>, ii: &mut Vec<usize>, l: usize, r: usize) {
            if l < r {
                let i_ = ii[l + rand_(r - l)];
                let mut i = l;
                let mut j = l;
                let mut k = r;
                while j < k {
                    let c = if aa[ii[j]] != aa[i_] { aa[ii[j]] - aa[i_] } else { ii[j] - i_ };
                    if c == 0 {
                        j += 1;
                    } else if c < 0 {
                        ii.swap(i, j);
                        i += 1;
                        j += 1;
                    } else {
                        k -= 1;
                        ii.swap(j, k);
                    }
                }
                sort(aa, ii, l, i);
                sort(aa, ii, k, r);
            }
        }

        sort(&aa, &mut ii, 0, n + 1);

        while q > 0 {
            let l: usize = lines.next().unwrap().unwrap().parse().unwrap() - 1;
            let r: usize = lines.next().unwrap().unwrap().parse().unwrap();

            let mut lower = -1;
            let mut upper = n + 1;

            while upper - lower > 1 {
                let h = (lower + upper) / 2;
                if aa[ii[h]] < aa[r] || (aa[ii[h]] == aa[r] && ii[h] < l) {
                    lower = h;
                } else {
                    upper = h;
                }
            }

            println!("{}", if aa[l] == aa[r] || (upper < n && ii[upper + 1] < ii[lower + 1]) { "YES" } else { "NO" });
            q -= 1;
        }

        t -= 1;
    }
}

fn rand_(n: usize) -> usize {
    use rand::Rng;
    rand::thread_rng().gen_range(0..n)
}

fn srand_() {
    use rand::Rng;
    let mut rng = rand::thread_rng();
    rng.gen();
}

File: ga/4omini_result_2024_7/result/1968F_mini/10_1968F.rs
use std::io::{self, Read};
use std::time::SystemTime;

const N: usize = 200000;

struct Random {
    z: u32,
}

impl Random {
    fn new() -> Self {
        let now = SystemTime::now();
        let time = now.duration_since(SystemTime::UNIX_EPOCH).expect("Time went backwards");
        let sec = time.as_secs() as u32;
        let usec = time.subsec_micros() as u32;
        let z = (sec ^ usec) | 1; // Ensure non-zero
        Random { z }
    }

    fn rand(&mut self) -> u32 {
        self.z *= 3;
        self.z >> 1
    }
}

fn sort(ii: &mut [usize], aa: &[u32], l: usize, r: usize, rand: &mut Random) {
    let mut l = l;
    let mut r = r;

    while l < r {
        let i = l;
        let j = l;
        let k = r;
        let i_ = ii[l + (rand.rand() % (r - l) as usize)];
        
        let mut j = j;
        let mut k = k;

        while j < k {
            let c = if aa[ii[j]] != aa[i_] {
                aa[ii[j]] as i32 - aa[i_] as i32
            } else {
                ii[j] as i32 - ii[i] as i32
            };

            if c == 0 {
                j += 1;
            } else if c < 0 {
                ii.swap(i, j);
                j += 1;
                l += 1;
            } else {
                k -= 1;
                ii.swap(j, k);
            }
        }
        sort(ii, aa, i, l, rand);
        l = k;
    }
}

fn search(aa: &[u32], ii: &[usize], a: u32, n: usize, i: usize) -> isize {
    let mut lower = -1;
    let mut upper = n as isize + 1;

    while upper - lower > 1 {
        let h = (lower + upper) / 2;

        if aa[ii[h as usize]] < a || (aa[ii[h as usize]] == a && ii[h as usize] < i) {
            lower = h;
        } else {
            upper = h;
        }
    }
    lower
}

fn main() {
    let input = io::stdin();
    let mut reader = input.lock();
    let mut buf = String::new();
    reader.read_to_string(&mut buf).unwrap();
    let mut iter = buf.split_whitespace();
    let mut rand = Random::new();

    let t: usize = iter.next().unwrap().parse().unwrap();
    for _ in 0..t {
        let n: usize = iter.next().unwrap().parse().unwrap();
        let q: usize = iter.next().unwrap().parse().unwrap();

        let mut aa = vec![0; n + 1];
        for i in 1..=n {
            let mut value: u32 = iter.next().unwrap().parse().unwrap();
            aa[i] = value ^ aa[i - 1];
        }

        let mut ii: Vec<usize> = (0..=n).collect();
        sort(&mut ii, &aa, 0, n + 1, &mut rand);

        for _ in 0..q {
            let l: usize = iter.next().unwrap().parse().unwrap();
            let r: usize = iter.next().unwrap().parse().unwrap();
            let l = l - 1;

            let result = if aa[l] == aa[r] || ii[search(&aa, &ii, aa[r], n, l) as usize + 1] < ii[search(&aa, &ii, aa[l], n, r) as usize] {
                "YES"
            } else {
                "NO"
            };
            println!("{}", result);
        }
    }
}